R/exposure_map_chart.R

```diff
@@ -0,0 +1,111 @@
+#' Create an interactive exposure map chart in an htmlwidget
+#'
+#' @param .data an exposure map data frame
+#' @param width width of exported htmlwidget in pixels (single integer value; default == NULL)
+#' @param height height of exported htmlwidget in pixels (single integer value; default == NULL)
+#' @param ... other options passed on to r2d3::r2d3() (see details)
+#'
+#' @description
+#' The `exposure_map_chart` function creates an interactive exposure map chart in an htmlwidget
+#'
+#' @md
+#' @export
+
+exposure_map_chart <-
+  function(.data, width = NULL, height = NULL, ...) {
+    .data <- export_data_utf8(.data)
+
+    dependencies <-
+      list(
+        system.file("js/map.js", package = "r2dii.interactive"),
+        system.file("js/d3-geo-projection.v2.min.js", package = "r2dii.interactive"),
+        system.file("js/d3-array.v2.min.js", package = "r2dii.interactive"),
+        system.file("js/world.js", package = "r2dii.interactive"),
+        system.file("js/text_dropdown_jiggle.js", package = "r2dii.interactive"),
+        system.file("css/2dii_gitbook_style.css", package = "r2dii.interactive"),
+        system.file("css/hide_styles.css", package = "r2dii.interactive")
+        )
+
+    op <- options(r2d3.shadow = FALSE)
+    on.exit(options(op), add = TRUE)
+
+    r2d3::r2d3(
+      data = .data,
+      script = system.file("render_exposure_map.js", package = "r2dii.interactive"),
+      dependencies = dependencies,
+      d3_version = 4,
+      width = width,
+      height = height,
+      container = "div",
+      ...
+    )
+  }
+
+
+#' Convert raw data into a exposure map data frame
+#'
+#' @param results_map results_map
+#' @param start_year start_year
+#' @param dataframe_translations dataframe_translations
+#' @param language_select two letter code for language (single string; default = "EN")
+#'
+#' @description
+#' The `as_exposure_map_data` function converts raw data into a exposure map data frame
+#'
+#' @import dplyr
+#' @export
+
+as_exposure_map_data <-
+  function(
+    results_map,
+    start_year,
+    dataframe_translations,
+    language_select = "EN"
+  ) {
```

Comment:
A someone who holds no preexisting context, I enjoy higher level abstractions that -- at a glance -- communicate the intent of the main function. Instead what I see is lower level details that I could understand but would take me quite a bit to grasp. In technical jargon, what I see here seems to be at a too low-level of abstraction. But I invite opposite views, maybe from @MonikaFu or @AlexAxthelm as external reviewers.

Reply:
What are you referring to? The arguments `start_year`, `dataframe_translations`, `language_select`?

Reply:
Sorry the placement of my comment is indeed confusing. I refer to the body of the function. I was suggesting it'd be nice for the the function to expose higher level abstractions named to communicate the main processes that are happening inside. Someting like

```r
f <- function(...) {
  check_inputs(...)

  if (something_happens()) {
    return(early)
  }
  
  do_main_thing()
}
```

This is a silly example that has nothing to do with your code, but I hope clarifies what I mean. The main function reads like the lead paragraph of a newspaper article, pushing the details below.

I could make an effort to push a commit trying to illustrate what I mean with a concrete example. I disclose that writing a good "lead" is super hard and I often don't do it because I run out of energy -- but I think it's a good thing to aim for.

Reply:
It's frustrating that we butt heads so often on these things... I very much dislike abstractions that force you, in order to understand what is happening, to jump to a different part of a file or even a completely different file. For me, that breaks my flow and pulls me out of the context I want to stay in. Obviously, there's a grey area where you weigh the pros and cons of doing one thing or the other... we tend to end up on the opposite sides of that. For me, I would never make a separate function to check the inputs of a parent function, which would not have any context accept directly inside the parent function... not to mention might have hilariously ridiculous consequences if you followed that to its logical end (check the inputs to that function, and check the inputs to that function, and check the inputs to that function, etc.).

What's happening in this function is...
1. a bit of setup: setting .data to NULL to work around dplyr's reliance on NSE and R CMD's obsessiveness (admittedly a hack), plus setting some default values for arguments, which normally I would do in the function definition, but you recently convinced me otherwise
2. code from create_interactive_report copy-pasted in with intentionally minimal changes

The code that came from create_interactive_report may be significantly improved and/or reduced in the future, but I'm not interested in putting a lot of work into that yet until I've tried this out a bit and also seen how other people use it or expect to use it before I make hard judgements about what the the best API would be. For example, I'm not sure filtering on asset class inside the function makes sense, and it may happen in the future that I completely remove that chunk of code and rather insist through the documentation that the user should filter that out themselves before sending it here, in which case the body of this function would become smaller and more easily digestible.

Reply:
I acknowledge the frustration and I thank you for your patience in keeping the dialog open. I feel this models how to have a conversation where two people hold to different believes while respecting opposing views. I hope this empowers other folks that may have several thoughts but stay silent.

Reply:
The reason I keep insisting with this "abstractions" is strongly influenced by what I read. But I may be interpreting what I read in an odd way, hence I would love to compare my interpretation of the literature against the interpretation of others, say @MonikaFu and @AlexAxthelm.

Reply:
Thanks. I appreciate your feedback even though I don't always agree with it (and often I do agree with it and/or I learn something new). I don't think there's a right or wrong answer to it. Both strategies have pros and cons. You and I weigh those differently (or maybe completely disregard some of them).

Reply:
In the end, I'm very okay with the maintainer to make the last call. It's them who understands  the code best than anyone else, and them the one who will enjoy or suffer the maintainability  if the project.

Reply:
or suffer the loss of potential collaborators (including their future self) who are flabbergasted by the enormous complexity of each simple task 😉

Reply:
this is pseudo-code to demonstrate what I imagine in my head when we have these debates....
``` r
do_a_thing <- function(...) {
  check_that_you_have_everything_you_need_to_do_the_thing(...)
  check_that_the_things_you_have_to_do_the_thing_are_what_you_expect(...)
  actually_do_the_thing(...)
}

check_that_you_have_everything_you_need_to_do_the_thing <- function(...) {
  check_if_things_are_missing(...)
}

check_that_the_things_you_have_to_do_the_thing_are_what_you_expect <- function() {
  ...
}

check_if_things_are_missing <- function(...) {
  lapply(..., check_if_a_thing_is_missing)
}

check_if_a_thing_is_missing <- function(x) {
  stop_with_message_if_missing(x, "x is missing")
}

stop_with_message_if_missing <- function(x, message) {
  stop_with_message_if_not(missing(x), message)
}

stop_with_message_if_not <- function(eval, message) {
  if (eval) { stop_with_message(message) }
}

stop_with_message <- function(message) {
  stop(message)
}

actually_do_the_thing <- function(x, y) {
  x + y
}
```

Reply:
Yeah, it's easy to fall into that rabbit hole. As Bob Martin says in Clean Code:

> It turns out to be very difficult for programmers to ... write functions that stay at a single level of abstraction. But learning this trick is also very important.

Maybe a discussion for the "practices" project or the devs-hangout.

--


![IMG-0081](https://user-images.githubusercontent.com/5856545/105427031-e1f41000-5c11-11eb-8c05-4678c34e34a1.jpg)

![IMG-0082](https://user-images.githubusercontent.com/5856545/105427024-de608900-5c11-11eb-8621-59ae1d17d857.jpg)

Reply:
I don't think that there necessarily is a right answer to this problem. From my experience - it can be frustrating, as @cjyetman mentions, to have to dig deep in order to find out what the function is actually doing. However, having the main functionalities wrapped into functions helps understanding what is happening in the function without having to read all the details of it, which can also be helpful depending on the willingness and need of the user to understand what the function is actually doing.

I'd say that in this case I would not go too far into refactoring before we decide how to refactor the `create_interactive_report` code and the `.js` charts, which are both on my TODO list. I believe that for now this package is even more WIP than `r2dii.ggplot` is and it might change in substantial ways so I am not sure if it makes sense to go deep into these kind of conversations at this stage.

Reply:
@maurolepore I'm going to resolve this conversation and merge this PR, but maybe we should move this discussion over to the Discussions in the best practices repo? I think it's interesting and maybe we get to the bottom of it eventually. I'm glad to hear that there are definitive limitations on the level of abstraction, which is not something I previously perceived in these discussions and is the root of at least one of my knee-jerk aversions to this type of suggestion.

Reply:
Sounds great; see https://github.com/2DegreesInvesting/practices/discussions/10.

I have a final message for those "watching" this PR because I realize it may be interpreted negatively and discourage contributions. Quite the opposite, CJ and I talk a lot about "how we talk" and seem to agree that it's okay to discuss  opposing views as long as they are respectful.

As we are trying to define good dialog, here I expose the system I used (inspired by [Crucial Conversations](https://youtu.be/ryCeARRE-KA)): I tried to first see where we agree (e.g. helpers sometimes obscure interesting information), then notice our common goals (e.g. write clear, maintainable code), and assume we only differ in the strategy we use to meet that goal (e.g. leave a code chunk in the main function or extract it into a helper).

We had a super friendly call just now and decided next time we'll move the conversation more quickly to the new GitHub tab "Discussions". That way we hope "watchers" won't need to read so many messages. Apologies for that :smile: